source/develop/release-management/features/storage/incremental-backup.html.md

```diff
@@ -28,138 +25,246 @@ allow incremental backup to start.
-ovirt-imageio via HTTPS.
-10. Ovirt-imageio obtain the changed blocks using qemu-img map and
-the external image.
-11. Backup application downloads the changed blocks (guest blocks)
-from ovirt-imageio via HTTPS.
-12. ovirt-imageio read the blocks from /dev/nbdX.
-
-
-## Preparing stopped VM disks for incremental backup
+   bitmap, and creating new dirty bitmap.
+1. oVirt queries the disk size needed for a backup operation, depending
+   on the dirty bitmap included in the backup.
+1. oVirt creates a temporary disk for copying changed blocks.  The
+   backing file of the temporary disk is the backed up snapshot.
+1. oVirt start backup operation with the external disk using pull mode.
+1. Qemu watches writes to the active snapshot. If blocks included in the
+   a the backup are modified, the blocks are copied to the temporary
+   disk to the backup will copy the correct data.
+1. oVirt starts local qemu-nbd, mounting the external disk at /dev/nbdX.
+1. oVirt notify the user (e.g. backup application) that the backup disk
+   is ready.
+1. Backup application starts a transfer with the relevant disks
+1. Backup application obtains the changed blocks list from ovirt-imageio
+   via HTTPS.
+1. Ovirt-imageio obtain the changed blocks using qemu-img map and the
+   external image.
+1. Backup application downloads the changed blocks (guest blocks) from
+   ovirt-imageio via HTTPS.
+1. ovirt-imageio read the blocks from /dev/nbdX.
+
+
+### Preparing stopped VM disks for incremental backup
+
+XXX needs update
+XXX need to create a new checkpoint
+
 This does not depend on the push or pull modes, but we need qemu
 support for getting the changed blocks from an image.
 
 Basically this is the same as push mode, after qemu pushed the
 changed blocks to external image.
 
 1. oVirt starts local qemu-nbd, mounting the external disk at /dev/nbdX.
-2. oVirt notify the user (e.g. backup application) that the backup
-disk is ready.
-3. Backup application starts a transfer with the relevant disks
-4. Backup application obtains the changed blocks list from
-ovirt-imageio via HTTPS.
-5. ovirt-imageio obtains the changed blocks using qemu-img map and
-the external image.
-6. Backup application downloads the changed blocks (guest blocks)
-from ovirt-imageio via HTTPS.
-7. ovirt-imageio reads the blocks from /dev/nbdX.
+1. oVirt notify the user (e.g. backup application) that the backup disk
+   is ready.
+1. Backup application starts a transfer with the relevant disks
+1. Backup application obtains the changed blocks list from ovirt-imageio
+   via HTTPS.
+1. ovirt-imageio obtains the changed blocks using qemu-img map and the
+   external image.
+1. Backup application downloads the changed blocks (guest blocks) from
+   ovirt-imageio via HTTPS.
+1. ovirt-imageio reads the blocks from /dev/nbdX.
+
+
+### Finalizing backup
+
+When backup succeeds, backup application completed to copy all data.
+
+1. Delete previous backup checkpoint using libvirt API
```

Comment:
Cleaning up old checkpoints must be configurable, since backup application
may need to have several independent jobs creating checkpoints. Deleting
checkpoints in one job may break the other job on the next incremental backup.

Same issue exist if you have 2 backup applications, for example switching 
from one solution to another. In the transition duration, you may backup
twice with both solutions, and both backups should not interfere with each other.

I think what can work is:
- make checkpoint cleanup optional
- allow the backup application to determine which checkpoints should be kept

Here is an example flow:
- hourly job create backup (and checkpoint) every hour
  will not clean up checkpoints
- daily job create backup (and checkpoint) every day
  cleanup all checkpoints of the previous day.

In this configuration, we have 24-48 checkpoints in an image.